src/common/subscribe.ts

```diff
@@ -0,0 +1,269 @@
+            })
+        })
+    )
+
+    try {
+        const user = (responses[0] || {}) as Record<string, any>
+        const models = (responses[1] || []) as OpenAIModelDetail[]
+        const authCode: string = user?.authCode || ''
+
+        if (!authCode || typeof user.authCode !== 'string') {
+            return undefined
+        }
+
+        let model: OpenAIModelDetail = defaultOpenAIModelDetail
+        if (models) {
+            const modelIndexes = new Map<string, OpenAIModelDetail>()
+            for (let index = 0; index < models.length; index++) {
+                const item = models[index]
+                if (!item?.id) {
+                    continue
+                }
+                modelIndexes.set(item.id, item)
+            }
+
+            model = modelIndexes.get('gpt-4') || modelIndexes.get(defaultAPIModel) || models[0]
+        }
+
+        return { authCode: authCode, model: model }
+    } catch {
+        return undefined
+    }
+}
+
+async function persistAuthSettings(settings: ISettings, key: string, value: AuthCodeModel | undefined): Promise<void> {
+    if (!settings || !key || !settings.providersProps['ThirdPartyChatGPT']) {
+        return
+    }
+
+    const extra: Record<string, any> = settings.providersProps['ThirdPartyChatGPT'].extra || {}
+    const auths = new Map<string, AuthCodeModel>(Object.entries(extra.auths || {}))
+    if (value === undefined) {
+        // not exist or delete failed
+        if (!auths.get(key) || !auths.delete(key)) {
+            return
+        }
+    } else {
+        auths.set(key, value)
+    }
+
+    extra.auths = Object.fromEntries(auths)
+    settings.providersProps['ThirdPartyChatGPT'].extra = extra
+
+    // apply all changes to the original settings object before saving
+    Object.assign(settings.providersProps['ThirdPartyChatGPT'].extra, extra)
+    return setSettings(settings)
+}
+
+export class AuthCodeModelManager {
+    private static instance: AuthCodeModelManager
+    private settings: ISettings
+    private cache: LRUCache<string, AuthCodeModel>
+
+    private constructor(settings: ISettings) {
+        this.settings = settings
+        this.cache = new LRUCache<string, AuthCodeModel>({
+            max: 50,
+            maxSize: 500,
+            // eslint-disable-next-line @typescript-eslint/no-unused-vars
+            sizeCalculation: (_value, _key) => {
+                return 1
+            },
+        })
+
+        const auths = settings?.providersProps['ThirdPartyChatGPT']?.extra?.auths
+        try {
+            if (auths && typeof auths === 'object') {
+                Object.entries(auths).map(([k, v]) => {
+                    const authCodeModel: AuthCodeModel = v as AuthCodeModel
+                    if (k && authCodeModel.authCode) {
+                        this.cache.set(k, authCodeModel)
+                    }
+                })
+            }
+        } catch {
+            console.error(`cache auth code and model from settings error`)
```

Comment:
Will `this.cache.set` cause an error? Maybe we can remove the try-catch block.

Reply:
There should be no errors theoretically, and I will remove it later.